File: scripts/senso-train/vtrick_features.py
"""Python port of mcts/vtrick-features.ts — the expected-tier net's features from
collect-value.ts ply rows (the full dealt world). Pinned to the TS encoder by
`check_fixture`, run by the trainer before training.
"""
from __future__ import annotations

import json
from pathlib import Path

import numpy as np

N_CARDS = 54
SEATS, TIERS = 5, 5
OWNER_CLASSES = 7
TABLE_SLOTS, PER_SLOT = 4, 6
FEATURES = N_CARDS * OWNER_CLASSES + TABLE_SLOTS * PER_SLOT + SEATS + 5 + 1 + SEATS * 2 + SEATS * 4 + 4 + SEATS + 3

# Card index → suit (-1 Ninja) and rank (0 Ninja): FULL_DECK order = 4 clans × ranks 2..14, then wood, jade.
CARD_SUIT = np.array([c // 13 for c in range(52)] + [-1, -1])
CARD_RANK = np.array([2 + (c % 13) for c in range(52)] + [0, 0])
WOOD, JADE = 52, 53


def win_key(card: int, trump: int, lead: int) -> int:
    if card == JADE:
        return 300
    if card == WOOD:
        return 200
    suit = CARD_SUIT[card]
    if suit == trump:
        return 100 + int(CARD_RANK[card])
    if suit == lead:
        return int(CARD_RANK[card])
    return -1


def tier_index(tricks: np.ndarray) -> np.ndarray:
    out = np.zeros_like(tricks, dtype=int)
    out[tricks >= 1] = 1
    out[tricks >= 3] = 2
    out[tricks >= 5] = 3
    out[tricks >= 7] = 4
    return out


class PlyRows:
    def __init__(self, samples):
        self.s = samples
        self.cols = samples.columns
        self.at = {c: i for i, c in enumerate(self.cols)}
        self.features = FEATURES

    def block(self, name: str, k: int) -> np.ndarray:
        base = self.at[f"{name}[0]"]
        return self.s.data[:, base : base + k]

    def encode(self, rows: np.ndarray) -> np.ndarray:
        d = self.s.data
        m = len(rows)
        n = d[rows, self.at["n"]].astype(int)
        me = d[rows, self.at["seat"]].astype(int)
        leader = d[rows, self.at["leader"]].astype(int)
        trump = d[rows, self.at["trump"]].astype(int)
        table_len = d[rows, self.at["tableLen"]].astype(int)
        table = self.block("table", 5)[rows].astype(int)
        owner = self.block("owner", N_CARDS)[rows].astype(int)
        tricks = self.block("tricksWon", SEATS)[rows]
        hands = self.block("handSize", SEATS)[rows]
        voids = self.block("voids", SEATS)[rows].astype(int)
        rnd = d[rows, self.at["round"]]
        X = np.zeros((m, FEATURES), dtype=np.float32)
        idx = np.arange(m)
        k = 0
        for c in range(N_CARDS):
            o = owner[:, c]
            cls = np.where(o == -2, 5, np.where(o == -3, 6, (o - me + n) % n))
            X[idx, k + cls] = 1
            k += OWNER_CLASSES
        lead = np.where(table_len > 0, CARD_SUIT[np.clip(table[:, 0], 0, N_CARDS - 1)], -1)
        for i in range(TABLE_SLOTS):
            present = i < table_len
            card = np.clip(table[:, i], 0, N_CARDS - 1)
            X[:, k] = present
            X[:, k + 1] = np.where(present, CARD_RANK[card] / 14, 0)
            X[:, k + 2] = np.where(present & (CARD_SUIT[card] == trump), 1, 0)
            X[:, k + 3] = np.where(present & (CARD_SUIT[card] >= 0) & (CARD_SUIT[card] == lead), 1, 0)
            X[:, k + 4] = np.where(present & (card == WOOD), 1, 0)
            X[:, k + 5] = np.where(present & (card == JADE), 1, 0)
            k += PER_SLOT
        # current winner (relative)
        for r in range(m):
            if table_len[r] > 0:
                best, best_key = 0, -2
                for i in range(table_len[r]):
                    key = win_key(int(table[r, i]), int(trump[r]), int(lead[r]))
                    if key > best_key:
                        best, best_key = i, key
                winner = (leader[r] + best) % n[r]
                X[r, k + (winner - me[r] + n[r]) % n[r]] = 1
        k += SEATS
        X[idx, k + np.where(lead >= 0, lead, 4)] = 1
        k += 5
        X[:, k] = table_len / 4
        k += 1
        for r in range(SEATS):
            seat = (me + r) % n
            seated = r < n
            X[:, k + r] = np.where(seated, tricks[idx, seat] / 13, 0)
            X[:, k + SEATS + r] = np.where(seated, hands[idx, seat] / 13, 0)
        k += SEATS * 2
        for r in range(SEATS):
            seat = (me + r) % n
            seated = r < n
            for suit in range(4):
                X[:, k + r * 4 + suit] = np.where(seated, (voids[idx, seat] >> suit) & 1, 0)
        k += SEATS * 4
        X[idx, k + trump] = 1
        k += 4
        X[idx, k + (leader - me + n) % n] = 1
        k += SEATS
        cards_left = hands[idx, :].sum(axis=1)
        X[:, k] = cards_left / 50
        X[:, k + 1] = rnd / 8
        X[:, k + 2] = n / 5
        k += 3
        assert k == FEATURES, (k, FEATURES)
        return X

    def targets(self, rows: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
        """Tier index per RELATIVE seat (m, 5) and a seated mask (m, 5)."""
        d = self.s.data
        n = d[rows, self.at["n"]].astype(int)
        me = d[rows, self.at["seat"]].astype(int)
        final = self.block("finalTricks", SEATS)[rows]
        idx = np.arange(len(rows))
        y = np.zeros((len(rows), SEATS), dtype=int)
        mask = np.zeros((len(rows), SEATS), dtype=bool)
        for r in range(SEATS):
            seat = (me + r) % n
            seated = r < n
            y[:, r] = np.where(seated, tier_index(final[idx, seat]), 0)
            mask[:, r] = seated
        return y, mask


def check_fixture(columns: list[str], fixture_path: str | Path, atol: float = 1e-5) -> int:
    from samples import Samples

    fx = json.loads(Path(fixture_path).read_text())
    if fx["columns"] != columns:
        raise SystemExit("vtrick fixture: the ply columns changed since the fixture was dumped")
    pr = PlyRows(Samples("SNVT", 0, np.array(fx["rows"], dtype=np.float32), fx["columns"]))
    if fx["features"] != pr.features:
        raise SystemExit(f"vtrick features: TS has {fx['features']}, Python {pr.features}")
    rows = np.array([c["row"] for c in fx["cases"]])
    X = pr.encode(rows)
    expected = np.array([c["x"] for c in fx["cases"]], dtype=np.float32)
    worst = np.abs(X - expected).max()
    if worst > atol:
        bad = np.argwhere(np.abs(X - expected) > atol)[0]
        raise SystemExit(f"vtrick features diverge from the TS encoder: ply {bad[0]} feature {bad[1]} (|Δ| {worst:.3g})")
    return len(rows)

File: packages/core/src/games/senso-battle-for-japan/mcts/vtrick-features.ts
// Features of the expected-tier net (plan Phase 3): from a DETERMINIZED trick
// position (every card's owner known, as the search sees it inside one
// sampled deal) predict, per seat, the distribution over the tier that seat
// ends the round with. Its targets come from self-play where seats acted
// under their own information, so no double-dummy knowledge leaks into the
// prediction — and because the input is the dealt world, different deals give
// different values, which is what the paired PIMC needs to average over.
//
// Encoded relative to the seat to move (`f.turn`), one net for every chair;
// the Python port in scripts/senso-train/vtrick_features.py is pinned to this
// file by a fixture dumped from the collector's ply rows.

import {
  CARD_RANK,
  CARD_SUIT,
  createFastRound,
  type FastRound,
  JADE,
  N_CARDS,
  OUT,
  PLAYED,
  WOOD,
  winningIndex,
} from "./fast-round";

export const VTRICK_SEATS = 5;
export const VTRICK_TIERS = 5;
const OWNER_CLASSES = 7; // relative seat 0..4, played, undealt
const TABLE_SLOTS = 4;
const PER_SLOT = 6;
export const VTRICK_FEATURES =
  N_CARDS * OWNER_CLASSES +
  TABLE_SLOTS * PER_SLOT +
  VTRICK_SEATS + // current winner (relative)
  5 + // lead suit one-hot + none
  1 + // table length
  VTRICK_SEATS * 2 + // tricks won, hand size
  VTRICK_SEATS * 4 + // void mask
  4 + // trump
  VTRICK_SEATS + // leader (relative)
  3; // cards left, round, players

/**
 * Fill `out[0..VTRICK_FEATURES)` from the dealt world `f`, relative to `f.turn`.
 * `voidMask` = per seat, bit `suit` set when known void; `round` 1..8.
 */
export function vtrickFeatures(
  f: FastRound,
  voidMask: Uint8Array,
  round: number,
  out: Float32Array,
): void {
  const n = f.n;
  const me = f.turn;
  const rel = (seat: number) => (seat - me + n) % n;
  out.fill(0, 0, VTRICK_FEATURES);
  let k = 0;
  for (let c = 0; c < N_CARDS; c++) {
    const o = f.owner[c];
    const cls = o === PLAYED ? 5 : o === OUT ? 6 : rel(o);
    out[k + cls] = 1;
    k += OWNER_CLASSES;
  }
  for (let i = 0; i < TABLE_SLOTS; i++) {
    if (i < f.tableLen) {
      const card = f.table[i];
      out[k] = 1;
      out[k + 1] = CARD_RANK[card] / 14;
      out[k + 2] = CARD_SUIT[card] === f.trump ? 1 : 0;
      out[k + 3] = CARD_SUIT[card] >= 0 && CARD_SUIT[card] === f.leadSuit ? 1 : 0;
      out[k + 4] = card === WOOD ? 1 : 0;
      out[k + 5] = card === JADE ? 1 : 0;
    }
    k += PER_SLOT;
  }
  if (f.tableLen > 0) out[k + rel((f.leader + winningIndex(f)) % n)] = 1;
  k += VTRICK_SEATS;
  out[k + (f.leadSuit >= 0 ? f.leadSuit : 4)] = 1;
  k += 5;
  out[k++] = f.tableLen / 4;
  for (let r = 0; r < VTRICK_SEATS; r++) {
    const seat = (me + r) % n;
    out[k + r] = r < n ? f.tricksWon[seat] / 13 : 0;
    out[k + VTRICK_SEATS + r] = r < n ? f.handSize[seat] / 13 : 0;
  }
  k += VTRICK_SEATS * 2;
  for (let r = 0; r < VTRICK_SEATS; r++) {
    const seat = (me + r) % n;
    if (r < n)
      for (let suit = 0; suit < 4; suit++) out[k + r * 4 + suit] = (voidMask[seat] >> suit) & 1;
  }
  k += VTRICK_SEATS * 4;
  out[k + f.trump] = 1;
  k += 4;
  out[k + rel(f.leader)] = 1;
  k += VTRICK_SEATS;
  out[k++] = f.cardsLeft / 50;
  out[k++] = round / 8;
  out[k++] = n / 5;
  if (k !== VTRICK_FEATURES) throw new Error(`vtrick features: filled ${k} of ${VTRICK_FEATURES}`);
}

/** Tier index (0 = none, 1..4 = tiers 1/3/5/7) of a trick count. */
export function tierIndexOf(tricks: number): number {
  if (tricks >= 7) return 4;
  if (tricks >= 5) return 3;
  if (tricks >= 3) return 2;
  if (tricks >= 1) return 1;
  return 0;
}

/**
 * Softmax the net's logits (5 seats × 5 tiers, relative order) into
 * absolute-seat probabilities `probs[(seat * 5) + t]`; unseated rows stay 0.
 */
export function tierProbabilities(logits: Float32Array, f: FastRound, probs: Float32Array): void {
  const n = f.n;
  const me = f.turn;
  probs.fill(0);
  for (let r = 0; r < n; r++) {
    const seat = (me + r) % n;
    let max = Number.NEGATIVE_INFINITY;
    for (let t = 0; t < VTRICK_TIERS; t++) max = Math.max(max, logits[r * VTRICK_TIERS + t]);
    let sum = 0;
    for (let t = 0; t < VTRICK_TIERS; t++) {
      const e = Math.exp(logits[r * VTRICK_TIERS + t] - max);
      probs[seat * VTRICK_TIERS + t] = e;
      sum += e;
    }
    for (let t = 0; t < VTRICK_TIERS; t++) probs[seat * VTRICK_TIERS + t] /= sum;
  }
}

/** Column lookups over the collector's raw ply rows (collect-value.ts). */
export interface PlyRowColumns {
  index: (name: string) => number;
}

/** Rebuild the dealt world of a collector ply row (the fixture path; tests). */
export function fastFromPlyRow(row: ArrayLike<number>, at: PlyRowColumns): FastRound {
  const n = row[at.index("n")];
  const f = createFastRound(n);
  const owner0 = at.index("owner[0]");
  for (let c = 0; c < N_CARDS; c++) f.owner[c] = row[owner0 + c];
  f.trump = row[at.index("trump")];
  f.leader = row[at.index("leader")];
  f.turn = row[at.index("seat")];
  f.tableLen = row[at.index("tableLen")];
  const table0 = at.index("table[0]");
  for (let i = 0; i < f.tableLen; i++) f.table[i] = row[table0 + i];
  f.leadSuit = f.tableLen > 0 ? CARD_SUIT[f.table[0]] : -1;
  const tricks0 = at.index("tricksWon[0]");
  const hand0 = at.index("handSize[0]");
  f.cardsLeft = 0;
  for (let s = 0; s < n; s++) {
    f.tricksWon[s] = row[tricks0 + s];
    f.handSize[s] = row[hand0 + s];
    f.cardsLeft += f.handSize[s];
  }
  return f;
}
